panic in v1.7.0

### Describe the bug
Panic occured randomly in one of our microservices.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xe01628]

goroutine 327 [running]:
github.com/fsnotify/fsnotify.(*watches).remove(0xc0007bdec0, 0x90de68?)
        /go/pkg/mod/github.com/fsnotify/fsnotify@v1.7.0/backend_inotify.go:182 +0x88
github.com/fsnotify/fsnotify.(*Watcher).readEvents(0xc00012cd00)
        /go/pkg/mod/github.com/fsnotify/fsnotify@v1.7.0/backend_inotify.go:535 +0x4a5
created by github.com/fsnotify/fsnotify.NewBufferedWatcher
        /go/pkg/mod/github.com/fsnotify/fsnotify@v1.7.0/backend_inotify.go:270 +0x1f0
```

### Code to Reproduce

Not able to reproduce

### File operations to reproduce

NA

### Which operating system and version are you using?

Linux 6.1.58+ SMP PREEMPT_DYNAMIC Mon Jan 29 15:19:25 UTC 2024 x86_64 Linux

### Which fsnotify version are you using?

v1.7.0

### Did you try the latest main branch?

No

I just stumbled upon the same issue, and it reproduces consistently running one of the Cilium's unit tests a few times:

```
$ git init 
$ git remote add origin https://github.com/cilium/cilium.git
$ git fetch --depth 1 origin 715ed4c20572b354cc875a81463d636624da31d4
$ git checkout 715ed4c20572b354cc875a81463d636624da31d4

$ go test -c ./pkg/clustermesh/common
$ ./common.test --test.v --test.run Test/ClusterMeshTestSuite/TestWatchConfigDirectory --test.count=100 --test.failfast 

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x87b8f9]

goroutine 550 [running]:
github.com/fsnotify/fsnotify.(*watches).remove(0xc001089e00, 0x5)
        /tmp/tmp.Uqhjz3pRld/vendor/github.com/fsnotify/fsnotify/backend_inotify.go:182 +0x79
github.com/fsnotify/fsnotify.(*Watcher).readEvents(0xc000a32280)
        /tmp/tmp.Uqhjz3pRld/vendor/github.com/fsnotify/fsnotify/backend_inotify.go:535 +0x4d2
created by github.com/fsnotify/fsnotify.NewBufferedWatcher in goroutine 549
        /tmp/tmp.Uqhjz3pRld/vendor/github.com/fsnotify/fsnotify/backend_inotify.go:270 +0x1c5
```

Output updating fsnotify to 77d5fee912901a9a5c56997ea6443af403472c79 and running with `FSNOTIFY_DEBUG=1`

```
FSNOTIFY_DEBUG: 18:59:47.199009999  AddWith("/tmp/multicluster734350355")
FSNOTIFY_DEBUG: 18:59:47.199131429  AddWith("/tmp/multicluster734350355/cluster1")
FSNOTIFY_DEBUG: 18:59:47.199292685  AddWith("/tmp/multicluster734350355/cluster2")
FSNOTIFY_DEBUG: 18:59:47.199467407           4:IN_ATTRIB                      → "/tmp/multicluster734350355/cluster1"
FSNOTIFY_DEBUG: 18:59:47.199496820        1024:IN_DELETE_SELF                 → "/tmp/multicluster734350355/cluster1"
FSNOTIFY_DEBUG: 18:59:47.199567928  Remove("/tmp/multicluster734350355/cluster1")
FSNOTIFY_DEBUG: 18:59:47.199670613       32768:IN_IGNORED                     → ""
FSNOTIFY_DEBUG: 18:59:47.199687176         512:IN_DELETE                      → "/tmp/multicluster734350355/cluster1"
FSNOTIFY_DEBUG: 18:59:47.210065264         256:IN_CREATE                      → "/tmp/multicluster734350355/cluster3"
FSNOTIFY_DEBUG: 18:59:47.210237310  AddWith("/tmp/multicluster734350355/cluster3")
FSNOTIFY_DEBUG: 18:59:47.210256957           2:IN_MODIFY                      → "/tmp/multicluster734350355/cluster3"
FSNOTIFY_DEBUG: 18:59:47.220508585          64:IN_MOVED_FROM                  → "/tmp/multicluster734350355/cluster3"
FSNOTIFY_DEBUG: 18:59:47.220541191         128:IN_CLASSA_MAX | IN_MOVED_TO    → "/tmp/multicluster734350355/cluster1"
FSNOTIFY_DEBUG: 18:59:47.220582465  Remove("/tmp/multicluster734350355/cluster3")
FSNOTIFY_DEBUG: 18:59:47.220679483  AddWith("/tmp/multicluster734350355/cluster1")
FSNOTIFY_DEBUG: 18:59:47.220705535        2048:IN_MOVE_SELF                   → ""
FSNOTIFY_DEBUG: 18:59:47.220753704       32768:IN_IGNORED                     → ""
FSNOTIFY_DEBUG: 18:59:47.230774923           4:IN_ATTRIB                      → "/tmp/multicluster734350355/cluster1"
FSNOTIFY_DEBUG: 18:59:47.230794723           4:IN_ATTRIB                      → "/tmp/multicluster734350355/cluster1"
FSNOTIFY_DEBUG: 18:59:47.331501395         256:IN_CREATE                      → "/tmp/multicluster734350355/..data_tmp"
FSNOTIFY_DEBUG: 18:59:47.331588170          64:IN_MOVED_FROM                  → "/tmp/multicluster734350355/..data_tmp"
FSNOTIFY_DEBUG: 18:59:47.331731836         128:IN_CLASSA_MAX | IN_MOVED_TO    → "/tmp/multicluster734350355/..data"
FSNOTIFY_DEBUG: 18:59:47.331785980           4:IN_ATTRIB                      → "/tmp/multicluster734350355/cluster2"
FSNOTIFY_DEBUG: 18:59:47.331896182        1024:IN_DELETE_SELF                 → "/tmp/multicluster734350355/cluster2"
FSNOTIFY_DEBUG: 18:59:47.332115852  AddWith("/tmp/multicluster734350355/cluster2")
FSNOTIFY_DEBUG: 18:59:47.332231977       32768:IN_IGNORED                     → ""
FSNOTIFY_DEBUG: 18:59:47.332257640  1073742336:IN_DELETE | IN_ISDIR           → "/tmp/multicluster734350355/..data-1"
FSNOTIFY_DEBUG: 18:59:47.332482645         256:IN_CREATE                      → "/tmp/multicluster734350355/..data_tmp"
FSNOTIFY_DEBUG: 18:59:47.332537092          64:IN_MOVED_FROM                  → "/tmp/multicluster734350355/..data_tmp"
FSNOTIFY_DEBUG: 18:59:47.332599916         128:IN_CLASSA_MAX | IN_MOVED_TO    → "/tmp/multicluster734350355/..data"
FSNOTIFY_DEBUG: 18:59:47.332655741           4:IN_ATTRIB                      → "/tmp/multicluster734350355/cluster2"
FSNOTIFY_DEBUG: 18:59:47.332722224        1024:IN_DELETE_SELF                 → "/tmp/multicluster734350355/cluster2"
FSNOTIFY_DEBUG: 18:59:47.332757686  AddWith("/tmp/multicluster734350355/cluster2")
FSNOTIFY_DEBUG: 18:59:47.332860668       32768:IN_IGNORED                     → ""
FSNOTIFY_DEBUG: 18:59:47.332883458  1073742336:IN_DELETE | IN_ISDIR           → "/tmp/multicluster734350355/..data-2"
FSNOTIFY_DEBUG: 18:59:47.332905160           4:IN_ATTRIB                      → "/tmp/multicluster734350355/cluster1"
FSNOTIFY_DEBUG: 18:59:47.332965314  Remove("/tmp/multicluster734350355/cluster1")
FSNOTIFY_DEBUG: 18:59:47.332989726        1024:IN_DELETE_SELF                 → "/tmp/multicluster734350355/cluster1"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x87bc39]

goroutine 167 [running]:
github.com/fsnotify/fsnotify.(*watches).remove(0xc000c8a480, 0x5)
        /tmp/tmp.Uqhjz3pRld/vendor/github.com/fsnotify/fsnotify/backend_inotify.go:178 +0x79
github.com/fsnotify/fsnotify.(*Watcher).readEvents(0xc000aca000)
        /tmp/tmp.Uqhjz3pRld/vendor/github.com/fsnotify/fsnotify/backend_inotify.go:554 +0x5f8
created by github.com/fsnotify/fsnotify.NewBufferedWatcher in goroutine 166
        /tmp/tmp.Uqhjz3pRld/vendor/github.com/fsnotify/fsnotify/backend_inotify.go:266 +0x1c5
```

Feel free to let me know if you need more information for troubleshooting it.

Here's a basic test case:

```
func TestRemoveRace(t *testing.T) {
	tmp := t.TempDir()

	w := newCollector(t, tmp)
	w.collect(t)

	for i := 0; i < 100; i++ {
		go mkdirAll(t, tmp, "/dir1")
		go addWatch(t, w.w, tmp, "/dir1")
		go rmAll(t, tmp, "/dir1")
		go rmWatch(t, w.w, tmp, "/dir1")
	}

	time.Sleep(1 * time.Second)

	ev := w.stop(t)
	fmt.Println(ev)
}
```

```
[~cg/fsnotify](main)% go test -run TestRemoveRace
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x5438f8]

goroutine 19 [running]:
github.com/fsnotify/fsnotify.(*inotify).readEvents(0xc0000e8140)
        /home/martin/code/Golib/fsnotify/backend_inotify.go:524 +0x838
created by github.com/fsnotify/fsnotify.newBufferedBackend in goroutine 18
        /home/martin/code/Golib/fsnotify/backend_inotify.go:191 +0x196
exit status 2
FAIL    github.com/fsnotify/fsnotify    0.007s
```

--- a/backend_inotify.go
+++ b/backend_inotify.go
@@ -91,7 +91,11 @@ func (w *watches) add(ww *watch) {
 func (w *watches) remove(wd uint32) {
 	w.mu.Lock()
 	defer w.mu.Unlock()
-	delete(w.path, w.wd[wd].path)
+	watch := w.wd[wd] // Could have had Remove() called. See #616.
+	if watch == nil {
+		return
+	}
+	delete(w.path, watch.path)
 	delete(w.wd, wd)
 }
 
@@ -469,20 +473,26 @@ func (w *inotify) readEvents() {
 				}
 			}
 
-			// If the event happened to the watched directory or the watched file, the kernel
-			// doesn't append the filename to the event, but we would like to always fill the
-			// the "Name" field with a valid filename. We retrieve the path of the watch from
-			// the "paths" map.
+			/// If the event happened to the watched directory or the watched
+			/// file, the kernel doesn't append the filename to the event, but
+			/// we would like to always fill the the "Name" field with a valid
+			/// filename. We retrieve the path of the watch from the "paths"
+			/// map.
 			watch := w.watches.byWd(uint32(raw.Wd))
-
-			var name string
-			if watch != nil {
-				name = watch.path
+			/// Can be nil if Remove() was called in another goroutine for this
+			/// path inbetween reading the events from the kernel and reading
+			/// the internal state. Not much we can do about it, so just skip.
+			/// See #616.
+			if watch == nil {
+				next()
+				continue
 			}
+
+			name := watch.path
 			if nameLen > 0 {
-				// Point "bytes" at the first byte of the filename
+				/// Point "bytes" at the first byte of the filename
 				bytes := (*[unix.PathMax]byte)(unsafe.Pointer(&buf[offset+unix.SizeofInotifyEvent]))[:nameLen:nameLen]
-				// The filename is padded with NULL bytes. TrimRight() gets rid of those.
+				/// The filename is padded with NULL bytes. TrimRight() gets rid of those.
 				name += "/" + strings.TrimRight(string(bytes[0:nameLen]), "\000")
 			}
 
@@ -497,14 +507,14 @@ func (w *inotify) readEvents() {
 
 			// inotify will automatically remove the watch on deletes; just need
 			// to clean our state here.
-			if watch != nil && mask&unix.IN_DELETE_SELF == unix.IN_DELETE_SELF {
+			if mask&unix.IN_DELETE_SELF == unix.IN_DELETE_SELF {
 				w.watches.remove(watch.wd)
 			}
 
 			// We can't really update the state when a watched path is moved;
 			// only IN_MOVE_SELF is sent and not IN_MOVED_{FROM,TO}. So remove
 			// the watch.
-			if watch != nil && mask&unix.IN_MOVE_SELF == unix.IN_MOVE_SELF {
+			if mask&unix.IN_MOVE_SELF == unix.IN_MOVE_SELF {
 				if watch.recurse {
 					next() // Do nothing
 					continue
@@ -529,7 +539,7 @@ func (w *inotify) readEvents() {
 
 			ev := w.newEvent(name, mask, raw.Cookie)
 			// Need to update watch path for recurse.
-			if watch != nil && watch.recurse {
+			if watch.recurse {
 				isDir := mask&unix.IN_ISDIR == unix.IN_ISDIR
 				/// New directory created: set up watch on it.
 				if isDir && ev.Has(Create) {

--- a/fsnotify_test.go
+++ b/fsnotify_test.go
@@ -821,3 +821,22 @@ func BenchmarkAddRemove(b *testing.B) {
 		do(b, w)
 	})
 }
+
+// Would panic on inotify: https://github.com/fsnotify/fsnotify/issues/616
+func TestRemoveRace(t *testing.T) {
+	t.Parallel()
+
+	tmp := t.TempDir()
+	w := newCollector(t, tmp)
+	w.collect(t)
+
+	dir := join(tmp, "/dir")
+	for i := 0; i < 100; i++ {
+		go os.MkdirAll(dir, 0o0755)
+		go os.RemoveAll(dir)
+		go w.w.Add(dir)
+		go w.w.Remove(dir)
+	}
+	time.Sleep(100 * time.Millisecond)
+	w.stop(t)
+}

